compiler/src/dotty/tools/debug/ExtractExpression.scala

```diff
@@ -0,0 +1,322 @@
+
+  private def setLocalValue(tree: Tree)(variable: Symbol, rhs: Tree)(using Context): Tree =
+    val strategy = ReflectEvalStrategy.LocalValueAssign(variable.asTerm)
+    reflectEval(tree)(nullLiteral, strategy, List(rhs), tree.tpe)
+
+  private def getClassCapture(tree: Tree)(variable: Symbol, cls: ClassSymbol)(using Context): Tree =
+    val byName = isByNameParam(variable.info)
+    val strategy = ReflectEvalStrategy.ClassCapture(variable.asTerm, cls, byName)
+    val qualifier = thisOrOuterValue(tree)(cls)
+    reflectEval(tree)(qualifier, strategy, List.empty, tree.tpe)
+
+  private def setClassCapture(tree: Tree)(variable: Symbol, cls: ClassSymbol, value: Tree)(using Context) =
+    val strategy = ReflectEvalStrategy.ClassCaptureAssign(variable.asTerm, cls)
+    val qualifier = thisOrOuterValue(tree)(cls)
+    reflectEval(tree)(qualifier, strategy, List(value), tree.tpe)
+
+  private def getMethodCapture(tree: Tree)(variable: Symbol, method: TermSymbol)(using Context): Tree =
+    val isByName = isByNameParam(variable.info)
+    val strategy =
+      ReflectEvalStrategy.MethodCapture(variable.asTerm, method.asTerm, isByName)
+    reflectEval(tree)(nullLiteral, strategy, List.empty, tree.tpe)
+
+  private def setMethodCapture(tree: Tree)(variable: Symbol, method: Symbol, value: Tree)(using Context) =
+    val strategy =
+      ReflectEvalStrategy.MethodCaptureAssign(variable.asTerm, method.asTerm)
+    reflectEval(tree)(nullLiteral, strategy, List(value), tree.tpe)
+
+  private def getStaticObject(tree: Tree)(obj: Symbol)(using ctx: Context): Tree =
+    val strategy = ReflectEvalStrategy.StaticObject(obj.asClass)
+    reflectEval(tree)(nullLiteral, strategy, List.empty, obj.typeRef)
+
+  private def getField(tree: Tree)(qualifier: Tree, field: TermSymbol)(using Context): Tree =
+    val byName = isByNameParam(field.info)
+    val strategy = ReflectEvalStrategy.Field(field, byName)
+    reflectEval(tree)(qualifier, strategy, List.empty, tree.tpe)
+
+  private def setField(tree: Tree)(qualifier: Tree, field: TermSymbol, rhs: Tree)(using Context): Tree =
+    val strategy = ReflectEvalStrategy.FieldAssign(field)
+    reflectEval(tree)(qualifier, strategy, List(rhs), tree.tpe)
+
+  private def callMethod(tree: Tree)(qualifier: Tree, method: TermSymbol, args: List[Tree])(using Context): Tree =
+    val strategy = ReflectEvalStrategy.MethodCall(method)
+    reflectEval(tree)(qualifier, strategy, args, tree.tpe)
+
+  private def callConstructor(tree: Tree)(qualifier: Tree, ctr: TermSymbol, args: List[Tree])(using Context): Tree =
+    val strategy = ReflectEvalStrategy.ConstructorCall(ctr, ctr.owner.asClass)
+    reflectEval(tree)(qualifier, strategy, args, tree.tpe)
+
+  private def reflectEval(tree: Tree)(
+      qualifier: Tree,
+      strategy: ReflectEvalStrategy,
+      args: List[Tree],
+      tpe: Type
+  )(using Context): Tree =
+    val evalArgs = List(
+      qualifier,
+      Literal(Constant(strategy.toString)),
```

Comment:
IIUC, this is for debugging purposes only? Might be worth a comment.

Reply:
done